[14.0][IMP] account_asset_management: Add new asset type unit of activity

matthieusaison opened the pull request:
This type of asset is depreciated based on its usage rather than the duration of usage.

Commit:
adds new asset type : unit of activity

florian-dacosta commented on account_asset_management/models/account_asset.py:
```diff
@@ -293,6 +293,26 @@ class AccountAsset(models.Model):
+                    raise exceptions.ValidationError("Number of usage can't be negativ")
+
+    @api.depends("total_number_of_use")
+    def _compute_remaining_usage(self):
```
Add "depreciation_line_ids.quantity" in dependencies

florian-dacosta commented on account_asset_management/models/account_asset.py:
```diff
@@ -293,6 +293,26 @@ class AccountAsset(models.Model):
         period.""",
     )
+    total_number_of_use = fields.Integer()
+    remaining_usage = fields.Integer(compute="_compute_remaining_usage")
```
I would store it, I don't really see why we would not.

florian-dacosta commented on account_asset_management/models/account_asset.py:
```diff
@@ -532,8 +552,11 @@ def validate(self):
+                    not asset.depreciation_line_ids.filtered(
+                        lambda l: l.type != "create"
+                    )
+                    and asset.method != "unit-activity"
```
I think it maybe better to override `compute_depreciation_board` and by pass the whole method for method = unit-activity, because it could be called at other places, what do you think ?

florian-dacosta reviewed (commented):
Some small remarks
Also, I think you can hide the field "Accumulate missed depreciations" for this method, as it does not really make sense.

Commit:
review enhancement

Commit:
fr translation

Commit:
verify asset line chronology

Commit:
Fix and updates

- Allow to edit total number_of_use of asset kind unit of activity
- Allow to set total_number_of_use in asset profile
- Issue:
On the view of supplier invoice, you can create an asset by specifying an asset_profile.
the _onchange_asset_profile_id trigger had a weird behavior:
If you have several asset profile with the same account_id configured.
In UI, if we change asset profile, the account_id change according to the on change
and trigger a cascade calculation of asset profile to the account associated on the asset profile.
This force the value of the first asset profile in profile_id
Fix: do not set the account id if it is identical to the one selected

florian-dacosta commented on account_asset_management/models/account_asset.py:
```diff
@@ -293,7 +293,9 @@ class AccountAsset(models.Model):
-    total_number_of_use = fields.Integer(store=True)
+    total_number_of_use = fields.Integer(
+        compute="_compute_total_number_of_use", store=True
+    )
```
I would add a readonly=False, so it can still be editable.
I think the values coming from the profile are always kind of default value but can always be forced

matthieusaison replied:
ok, done

florian-dacosta commented on account_asset_management/models/account_asset.py:
```diff
@@ -325,7 +327,15 @@ def _compute_move_line_check(self):
             asset.move_line_check = bool(
                 asset.depreciation_line_ids.filtered("move_id")
+                and not asset.depreciation_line_ids.filtered(
+                    lambda l: l.type == "create"
```
Actually, I am not sure we can change this field because it seems normal not to be able to change the purchase value if it comes from the invoice.
Maybe you would need to had a field like "has_depreciation_accounting_entry" that check if at least one depreciation lines of type `depreciate` has an accounting entry, and use this new field for the readonly of `total_number_of_use` ?
And keep the existing `move_line_check` for the `purchase_value`

Also, this condition seems wrong, because it checks if there is a line of type create, it does not check if it is linked to an `account.move`.

florian-dacosta commented on account_asset_management/models/account_asset_profile.py:
```diff
@@ -169,6 +169,7 @@ class AccountAssetProfile(models.Model):
         "posted depreciation line will prompt the option to reverse the "
         "journal entry, instead of deleting them.",
     )
+    total_number_of_use = fields.Integer(store=True)
```
You don't need to add `store=True` because it is not a computed (nor related) field. It is always stored by default.

sebastienbeau committed:
account_asset_management: fix showing line date
